### PR TITLE
docs: document buildQueryExpression

### PR DESCRIPTION
**Pull request by samisayegh**
Added ts annotations and updated the samples. Also renamed `createExpressionBuilder` -> `buildQueryExpression`, because it felt more consistent when I was searching for the export.

@dlutzCoveo, tagging you on the PR since it is adding documentation. Feel free to make suggestions, or changes directly on the branch, as you prefer

### CODE REVIEW

File: packages/samples/headless-react/src/pages/SearchPage.tsx [hunk 1]
```diff
@@ -356,6 +356,34 @@ export class SearchPage extends Component {
     this.unsubscribeExecuteTrigger();
   }
 
+  private get messageExpression() {
+    return buildQueryExpression({
+      operator: 'and',
```
Comment (samisayegh) [inline packages/samples/headless-react/src/pages/SearchPage.tsx:361]: We could make `operator` optional with `and` as the default. It was the value used in all the samples. It's also technically not necessary to specify if only adding one expression, so it would reduce the verbosity in this case,

Flip argument is making it required makes it clear how expressions are going to be concatenated, and helps with discoverability 🤷

Comment (ThibodeauJF) [reply packages/samples/headless-react/src/pages/SearchPage.tsx:361]: I would :+1: using `and` as by default

**Review by ThibodeauJF — APPROVED** [review]

File: packages/samples/headless-react/src/pages/SearchPage.tsx [hunk 2]
```diff
@@ -356,6 +356,34 @@ export class SearchPage extends Component {
...
+        operator: 'isExactly',
+        values: ['Confluence2Crawler'],
+      })
+      .addStringField({
+        field: 'documenttype',
+        operator: 'isExactly',
+        values: ['Space'],
+        negate: true,
+      })
+      .toString();
```
Comment (samisayegh) [inline packages/samples/headless-react/src/pages/SearchPage.tsx:384]: So, this is what building strings with the builder pattern looks like.

I was discussing with @acote-coveo how to make it less verbose, and he suggested having multiple composable functions/classes. Using such a design, the above expression might look something like:

```
const {and, stringField, not} = buildQueryExpression();

and(
   stringField({field: 'connectortype', operator: 'isExactly', values: ['Confluence2Crawler']}),
   not(stringField({field: 'documenttype', operator: 'isExactly', values: ['Space']}))
)
```

We could also remove the objects
```
and(
   stringField('connectortype', 'isExactly', ['Confluence2Crawler']),
   not(stringField('documenttype', 'isExactly', ['Space']))
)
```

We can do the same with the builder pattern too.

One drawback of not using objects is if there are ever multiple optional properties (e.g. other than `negate`), then one would need to specify them all to specify the optional arg they care about. e.g.

```
stringField(field, operator, value, _something optional_, the optional thing I want to specify).
```


For this, @acote-coveo suggested using overloads to allow both positional args and a single object. Typescript allows this e.g.

```
export function buildStringField(field:string, operator: StringOperator, values: string[]): Part;
export function buildStringField(expression: StringFieldExpression): Part;

export function buildStringField(...args: [StringFieldExpression] | [string, StringOperator, string[]]): Part {
  const config: StringFieldExpression = args.length === 1 ? args[0] : {field: args[0], operator: args[1], values: args[2]}
  return {
      ...
  }
})
```

For docs, the first overload is what will appear in the parsed json.


So, to sum up, there doesn't seem to be any technical limitations with the composed approach.
Question is, which api would you enjoy using more? Builder function, or multiple composable functions?

File: packages/headless/src/utils/query-expression/exact-match/exact-match.ts [hunk 3]
```diff
@@ -2,6 +2,9 @@ import {getNegationPrefix, Negatable} from '../common/negatable';
 import {Part} from '../common/part';
 
 export interface ExactMatchExpression extends Negatable {
+  /**
+   * An expression that must appear in its entirety at least once for an item to be returned.
+   */
```
Comment (btaillon) [inline packages/headless/src/utils/query-expression/exact-match/exact-match.ts:7]: I think we should give examples in this and `KeywordExpression` to help differentiate them.

File: packages/headless/src/utils/query-expression/date-field/date-field.ts [hunk 4]
```diff
@@ -3,8 +3,19 @@ import {getOperatorSymbol, NumericOperator} from '../common/operator';
...
 export interface DateFieldExpression extends Negatable {
+  /**
+   * The field name.
+   */
   field: string;
+
+  /**
+   * The operator to use to compare the `field` and `value`.
+   */
   operator: NumericOperator;
```
Comment (btaillon) [inline packages/headless/src/utils/query-expression/date-field/date-field.ts:14]: If there's doc for it, I think we should link doc in every `operator` parameter.

File: packages/headless/src/utils/query-expression/near/near.ts [hunk 5]
```diff
@@ -2,12 +2,26 @@ import {getNegationPrefix, Negatable} from '../common/negatable';
...
 export interface NearExpression extends Negatable {
+  /**
+   * The reference term.
+   */
   startTerm: string;
+
+  /**
+   * The other terms to check against the reference term.
+   */
   otherTerms: OtherTerm[];
```
Comment (btaillon) [inline packages/headless/src/utils/query-expression/near/near.ts:13]: I think we should give a "story" sort of example in those.

I'm not familiar with how those epxressions work, but I would think something like:
> If somebody wanted to search for a tempered glass screen protector for their phone but wanted to match both "iPhone screen protector (tempered glass)" and "tempered glass screen protector for iPhone", they could use:
> `{ startTerm: "screen protector", ["iPhone", "tempered glass"] }`

Don't use my exact example without checking with doc and search API though.

File: packages/samples/headless-react/src/pages/SearchPage.tsx (hunk 1 above)
Comment (btaillon) [reply packages/samples/headless-react/src/pages/SearchPage.tsx:361]: I think it's fine to use `and` as the default. My only worry is that, since this is a builder, some people may be looking for a `withOperator` method instead of a parameter.

File: packages/samples/headless-react/src/pages/SearchPage.tsx (hunk 2 above)
Comment (btaillon) [reply packages/samples/headless-react/src/pages/SearchPage.tsx:384]: Of course, a part of me feels it's wrong to see such long query expressions, but I don't think it's actually a bad thing. If somebody's expression contains too many terms where it gets hard to read, they can extract some parts in other variables or functions easily.

I'm more worried about potential issues with using helper methods.

**Review by btaillon — APPROVED** [review]

**Review by dlutzCoveo — APPROVED** [review]
Looks great!

File: packages/samples/headless-react/src/pages/SearchPage.tsx (hunk 1 above)
Comment (samisayegh) [reply packages/samples/headless-react/src/pages/SearchPage.tsx:361]: True, a method would be more aligned with the builder pattern, will adjust to that

File: packages/samples/headless-react/src/pages/SearchPage.tsx (hunk 2 above)
Comment (olamothe) [reply packages/samples/headless-react/src/pages/SearchPage.tsx:384]: Personally, I much prefer the builder pattern.

The main reason is that it's a very "classic" pattern from Java, and as such, if I was a user, I would instantly/naturally understand how to use it.

It's verbose and boring yeah, but most importantly it's * very * simple to understand. Boring is good.

It reads top to bottom, and I instantly know very well what I'll end up with at the end when calling `.toString()`

Versus this: 

```
and(
   stringField({field: 'connectortype', operator: 'isExactly', values: ['Confluence2Crawler']}),
   not(stringField({field: 'documenttype', operator: 'isExactly', values: ['Space']}))
)
```

I'm genuinely concerned that people would not actually understand how to use this system on their own without some documentation or copy pasting examples.

Comment (samisayegh) [reply packages/samples/headless-react/src/pages/SearchPage.tsx:384]: Yea, api design is tough. Personally I found multiple small functions was easier to follow, but it's also less conventional. It seems like the team prefers the builder though, so we'll continue as is,

File: packages/headless/src/utils/query-expression/near/near.ts (hunk 5 above)
Comment (samisayegh) [reply packages/headless/src/utils/query-expression/near/near.ts:13]: Ok, I think I will just to link to the [example in the documentation](https://docs.coveo.com/en/1814/searching-with-coveo/search-prefixes-and-operators#near) to help someone to learn more. Thanks,
